refactor(core): add `whenStable` private API

alan-agius4 opened the pull request:
Prior to this change `this.isStable.pipe(first((isStable) => isStable)).toPromise()` had to be done in multiple places across the framework and the Angular CLI see https://github.com/angular/angular-cli/pull/25856#discussion_r1328158846. In the majority of cases an Observable based `isStable` API is not needed. This also removes the need for RXJS operator imports.

ajafff commented on packages/core/src/application_ref.ts:
```diff
@@ -852,6 +852,16 @@ export class ApplicationRef {
 
+  private _whenStable: Promise<void>|undefined;
+  /**
+   * Returns a Promise that resolves when the application becomes first stable.
```
`ComponentFixture#whenStable()` resolves when it becomes stable the next time after the method is called. it's unfortunate that both methods have the same signature but work differently.

btw. "first stable" is not entirely true. because of the lazy initialization below, it should read as "when the application becomes stable after this method is called the first time."

ajafff commented on packages/core/src/application_ref.ts:
```diff
@@ -852,6 +852,16 @@ export class ApplicationRef {
+  /**
+   * Returns a Promise that resolves when the application becomes first stable.
+   */
+  get whenStable(): Promise<void> {
```
this is not tree-shakeable, so every application pays for it (and the imported rxjs operators) even if it doesn't use SSR or hydration

alan-agius4 replied:
That is true.

clydin replied:
`first` is already present in a new v17 application (along with a variety of other operators).

JeanMeche replied:
If this is the case, why are the symbol files changed in this PR ? It seems like it's adding `first()` and other functions it's pulling.

alan-agius4 replied:
This is because the symbols tests to do import all the modules in a `ng new` application. They only import the bare minimum of modules and functionality needed.

See: https://github.com/angular/angular/blob/main/packages/core/test/bundling/hello_world/index.ts

So indeed, while `first` looks like has been added to the application, in a real world application that use the router, it is already present.

Totati replied:
What about the deprecation of [`toPromise`](https://rxjs.dev/deprecations/to-promise) in favour of [`firstValueFrom`](https://rxjs.dev/api/index/function/firstValueFrom)?

JeanMeche replied:
The repo still maintains to compatibility with Rx6. So `firstValueFrom` isn't available.

Totati replied:
But I see that `firstValueFrom` is not used in the codebase.

JoostK commented on packages/core/src/application_ref.ts:
```diff
@@ -852,6 +852,17 @@ export class ApplicationRef {
+   */
+  get whenStable(): Promise<void> {
+    return (
+        this._whenStable ??=
+            this.isStable.pipe(first((isStable) => isStable)).toPromise().then(() => void 0));
```
I don't think this actually works as intended; caching the single `Promise` here means that once it's been resolved, it'll stay resolved forever whereas the stability of an app may toggle. The `toPromise` call drives a single subscription that'll complete once, then never again. Either the cache has to be cleared when the app becomes stable, but it'd be simpler to just drop the cached promise altogether.

Also, an interesting behavior difference with replacement APIs like `firstValueFrom` is their handling of a complete signal before a value has been emitted. `toPromise` would resolve to `undefined`, whereas the new APIs reject the promise with an `EmptyError`. I haven't looked at the tests whether such behavior would be caught by tests, if there's ever a refactor away from `toPromise`.

JeanMeche replied:
`first` throws `EmptyError` if nothing is emitted. Wouldn't that handle the case of no value ?

JoostK replied:
Oh good point, that avoids that particular difference in behavior :+1:

alan-agius4 replied:
> I don't think this actually works as intended; caching the single `Promise` here means that once it's been resolved, it'll stay resolved forever whereas the stability of an app may toggle. 

This is actually the intended behaviour once the application is stable for the first time rendering, hydration, cache disabled among other things are performed.

alan-agius4 replied:
I would kinda argue that currently it is not working as expected. Since we don't have a single source of truth when the app is first stable and ready to perform all the tasks at hand.

AndrewKushnir reviewed (commented):
I like the idea of having a unified function to avoid duplicated code and minimize the chance of introducing errors. I'm not sure though about exposing it publicly without a strong signal. We'd probably need some "when stable" API for zoneless as well and I'd propose to avoid adding extra public APIs until we have a chance to figure out the shape of that zoneless API.

My proposal is to create a standalone `whenStable` utility function, use it in the existing code and export as a private symbol for use in the CLI code.

alan-agius4 committed:
refactor(core): add `whenStable` private API

Prior to this change `this.isStable.pipe(first((isStable) => isStable)).toPromise()` had to be done in multiple places across the framework and the Angular CLI see https://github.com/angular/angular-cli/pull/25856#discussion_r1328158846. In the majority of cases an Observable based `isStable` API is not needed. This also removes the need for RXJS operator imports.

alan-agius4 commented on packages/core/src/hydration/api.ts:
```diff
@@ -184,7 +182,7 @@ export function withDomHydration(): EnvironmentProviders {
             // Note: the cleanup task *MUST* be scheduled within the Angular zone
             // to ensure that change detection is properly run afterward.
-            whenStable(appRef, injector).then(() => {
+            whenStableWithTimeout(appRef, injector).then(() => {
```
This is interesting, maybe we should include this logic as part of the `whenStable`. I can see that for SSR/SSG we'd might want to have a timeout to avoid problematic apps that never go stable from hanging during rendering on the server. Maybe for now have a sensible default and afterwards we expose it via a DI token that can be configured without direct access of this method.

Context: https://github.com/angular/angular/issues/51549

AndrewKushnir replied:
Yeah, that's a great idea 👍 If application remains unstable after a timeout, we reject the promise?

In this case we can handle both states:
```
whenStable(appRef, 1000 /* optional stabilization timeout */)
  .then(() => { /* app is stable */ })
  .catch(() => {
    if (ngDevMode) {
      console.log('This application remained unstable for a long time.');
    }
  });
```

clydin replied:
An `AbortSignal` may be an interesting option. It allows for the timeout as well as arbitrary user defined conditions via the abort controller. Plus we get `TimeoutError` and `AbortError` exceptions effectively for free.
https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal

alan-agius4 replied:
Let's add this in a followup!

alan-agius4 committed:
fixup! refactor(core): add `whenStable` private API

alfaproject commented on packages/core/src/application_ref.ts:
```diff
@@ -1301,3 +1301,25 @@ export function provideZoneChangeDetection(options?: NgZoneOptions): Environment
+  }
+
+  const whenStablePromise =
+      applicationRef.isStable.pipe(first((isStable) => isStable)).toPromise().then(() => void 0);
```
```suggestion
      applicationRef.isStable.pipe(first((isStable) => isStable)).toPromise().then(() => {});
```

alfaproject replied:
But, as long as this is private why even bother? The boolean value should be fine?

alan-agius4 replied:
So that the same logic can be used in the Angular CLI, expand on it, and eventually making it public. The boolean value is redundant in this case since it will never `false`.

Totati replied:
Isn't `applicationRef.isStable.pipe(first((isStable) => isStable)).toPromise()` enough? Like why the extra characters just to hide the `true` value?

alan-agius4 replied:
@Totati, those extra character will not make any difference in performance. IMHO such micro optimzations are really superfluous.

AndrewKushnir commented on packages/core/src/private_export.ts:
```diff
@@ -0,0 +1,9 @@
+/**
```
We can add the symbol to an existing `packages/core/src/core_private_export.ts` file.

alan-agius4 replied:
I was looking for that, but did not find it.

alan-agius4 committed:
fixup! refactor(core): add `whenStable` private API

alan-agius4 commented:
**Caretaker note:** I am unable to remove the unnecessary review request by @alxhub.

dylhunn commented:
I am also unable to remove it, and the caretaking tool won't allow me to merge it with active reviewers. @josephperrott mind trying to merge this one manually?
